File: src/test/java/im/tox/tox4j/callbacks/FriendNameCallbackTest.java
package im.tox.tox4j.callbacks;

import im.tox.tox4j.AliceBobTestBase;
import im.tox.tox4j.ToxCore;
import im.tox.tox4j.ToxCoreImpl;
import im.tox.tox4j.ToxOptions;
import im.tox.tox4j.annotations.NotNull;
import im.tox.tox4j.enums.ToxConnection;
import im.tox.tox4j.exceptions.ToxException;
import im.tox.tox4j.exceptions.ToxNewException;

import static org.junit.Assert.assertEquals;

public class FriendNameCallbackTest extends AliceBobTestBase {

    @NotNull
    @Override
    protected ChatClient newAlice() {
        return new Client();
    }


    private static class Client extends ChatClient {

        public void friendConnectionStatus(final int friendNumber, @NotNull ToxConnection connection) {
            if (connection != ToxConnection.NONE) {
                debug("is now connected to friend " + friendNumber);
                addTask(new Task() {
                    @Override
                    public void perform(@NotNull ToxCore tox) throws ToxException {
                        tox.setName(getName().getBytes());
                    }
                });
            }
        }

        @Override
        public void friendName(int friendNumber, @NotNull byte[] name) {
            debug("friend changed name to: " + new String(name));
            assertEquals(FRIEND_NUMBER, friendNumber);
            assertEquals(getFriendName(), new String(name));
            finish();
        }

    }

}

File: projects/tox4j/src/test/java/im/tox/tox4j/core/callbacks/FriendNameCallbackTest.scala
package im.tox.tox4j.core.callbacks

import im.tox.tox4j.core.enums.ToxConnection
import im.tox.tox4j.testing.autotest.{AliceBobTest, AliceBobTestBase}

final class FriendNameCallbackTest extends AliceBobTest {

  override type State = Int
  override def initialState: State = 0

  protected override def newChatClient(name: String, expectedFriendName: String) = new ChatClient(name, expectedFriendName) {

    override def friendConnectionStatus(friendNumber: Int, connectionStatus: ToxConnection)(state: ChatState): ChatState = {
      super.friendConnectionStatus(friendNumber, connectionStatus)(state)
      if (connectionStatus != ToxConnection.NONE) {
        state.addTask { (tox, state) =>
          tox.setName(selfName.getBytes)
          state
        }
      } else {
        state
      }
    }

    override def friendName(friendNumber: Int, name: Array[Byte])(state: ChatState): ChatState = {
      debug(s"friend changed name to: ${new String(name)}")
      assert(friendNumber == AliceBobTestBase.FriendNumber)

      state.get match {
        case 0 =>
          assert(name.isEmpty)
          state.set(1)
        case 1 =>
          assert(new String(name) == expectedFriendName)
          state.finish
      }
    }

  }

}
